### ISSUE TITLE
Rule proposal: Disallow implicit casts from/to `any`

### ISSUE THREAD

**Issue by emlai** [typescript-eslint/typescript-eslint#306]
In my experience a frequent source of errors in TypeScript is implicit conversion from/to `any`.

`noImplicitAny` and `no-explicit-any` do help, but sometimes third-party libraries return e.g. `any[]` that still needs to be cast. For example:

```ts
  const result = await database.query('SELECT * FROM foo')
  doSomethingWith(result.rows) // result.rows has type `any[]`
```

I propose to add a rule that flags the above implicit conversion from `result.rows` to whatever `doSomethingWith` parameter type is, and requires the user to cast it explicitly:

```ts
  doSomethingWith(result.rows as Foo[])
```

Now if the `doSomethingWith` parameter type changes, `result.rows` is not silently cast to a possibly wrong type.

**Comment by bradzacher** [comment]
I'm not sure how much value this has, at least from your example.
IMO this creates a "brittle" state in your code where you'll have to manually update an (unnecessary) typecast if the function signature ever changes.

If you've got some more examples to help justify it, I'd love to see them.

----

Just as a side note:

If you have control of the typings for the function (doubt you do, but as an FYI anyways), you can accomplish this with the `unknown` type:

```TS
declare var database: {
    query(q: string): {
        rows: unknown[],
    },
}

type Foo = string;
function doSomethingWith(rows: Foo[]) {}

const result = database.query('SELECT * FROM foo');
doSomethingWith(result.rows);
/*
Argument of type 'unknown[]' is not assignable to parameter of type 'string[]'.
  Type 'unknown' is not assignable to type 'string'.
*/
```
[repl](https://www.typescriptlang.org/play/index.html#src=declare%20var%20database%3A%20%7B%0D%0A%20%20%20%20query(q%3A%20string)%3A%20%7B%0D%0A%20%20%20%20%20%20%20%20rows%3A%20unknown%5B%5D%2C%0D%0A%20%20%20%20%7D%2C%0D%0A%7D%0D%0A%0D%0Atype%20Foo%20%3D%20string%3B%0D%0Afunction%20doSomethingWith(rows%3A%20Foo%5B%5D)%20%7B%7D%0D%0A%0D%0Aconst%20result%20%3D%20database.query('SELECT%20*%20FROM%20foo')%3B%0D%0AdoSomethingWith(result.rows)%3B%0D%0A)

**Comment by emlai** [comment]
> IMO this creates a "brittle" state in your code where you'll have to manually update an (unnecessary) typecast if the function signature ever changes.

I think you specifically do want to manually check the call site if the function signature changes, because `result.rows` might no longer have the correct type for the function.

**Comment by ThomasdenH** [comment]
Do you mean the [`no-unsafe-any` rule](https://palantir.github.io/tslint/rules/no-unsafe-any/)?

**Comment by ThomasdenH** [comment]
The `unknown` keyword is always preferred in a strict codebase. However, since many libraries do not yet use `unknown` for various reasons, the rule is essential for a statically checked project.

For example, the signature for `JSON.parse` should look like this: `(in: string) => unknown`. Instead, it looks like this: `(in: string): any`. This means that the following crashing code is allowed by Typescript:

```typescript
let a: { property: { subproperty: string } };
a = JSON.parse("{}");
console.log(a.property.subproperty);
```
[Playground](https://www.typescriptlang.org/play/#src=let%20a%3A%20%7B%20property%3A%20%7B%20subproperty%3A%20string%20%7D%20%7D%3B%0D%0Aa%20%3D%20JSON.parse(%22%7B%7D%22)%3B%0D%0Aconsole.log(a.property.subproperty)%3B%0D%0A)

The [proposal to add this option in Typescript](https://github.com/Microsoft/TypeScript/issues/22716) was closed, as the `unknown` type fixes this eventually, and the `no-unsafe-any` lint covers the other cases.

**Comment by mohsen1** [comment]
The TypeScript team [decided this is not useful as a strict flag](https://github.com/Microsoft/TypeScript/pull/24423). But I support this as an ESLint rule

**Comment by emlai** [comment]
`no-unsafe-any` [seems to allow](https://github.com/Microsoft/TypeScript/issues/22716#issuecomment-374638717) implicit conversions from `any[]` to `T[]`. I would like the lint rule to also catch those.

**Comment by ThomasdenH** [comment]
And preferably there would be a better error message that gives a hint about what coercion is happening.

**Comment by bradzacher** [comment]
I've written up a more detailed proposal #306 which covers more cases.
I'm closing this in favour of that.

**Comment by bradenhs** [comment]
>I've written up a more detailed proposal #306 which covers more cases.
I'm closing this in favour of that.

#306 is this issue

**Comment by bradzacher** [comment]
> #306 is this issue

![image](https://user-images.githubusercontent.com/7462525/62316505-6fefa000-b44c-11e9-9057-b57d0c244df7.png)


I meant - #791
